src/MicrosoftTeams.ts

```diff
@@ -421,22 +422,61 @@ namespace microsoftTeams
             let link = document.createElement("a");
             link.href = authenticateParameters.url;
 
-            let messageId = sendMessage(
-                "authentication.authenticate",
-                link.href,
-                authenticateParameters.width,
-                authenticateParameters.height);
-            callbacks[messageId] = (success: boolean, response: string) =>
+            // Clear any previous success/failure values
+            localStorage.removeItem("authentication.success");
+            localStorage.removeItem("authentication.failure");
+
+            // Open a popup window to the specified authentication URL
+            let authWindow: Window;
+            if (loadedInDesktopClient)
             {
-                if (success)
+                let messageId = sendMessage("authentication.authenticate", link.href);
+                callbacks[messageId] = (allowed: boolean) =>
                 {
-                    authenticateParameters.successCallback(response);
-                }
-                else
+                    if (allowed)
+                    {
+                        authWindow = currentWindow.open(link.href, "Microsoft Teams", "width = " + authenticateParameters.width + ", height = " + authenticateParameters.height);
+                    }
+                    else
+                    {
+                        throw new Error("Authentication is only supported for URLs matching the pattern registered in the manifest.");
+                    }
+                };
+            }
+            else
+            {
+                authWindow = currentWindow.open(link.href, "Microsoft Teams", "width = " + authenticateParameters.width + ", height = " + authenticateParameters.height);
+            }
+
+            if (!authWindow)
+            {
+                throw new Error("Failed to open authentication window.");
+            }
+
+            // Wait for the authentication window to close
+            let interval = setInterval(() =>
+            {
```

Comment:
Good catch; I thought checking the success/failure status was sufficient here but if the user closes the window before the authentication flow completes then we will keep spinning indefinitely. I added a check for the authWindow.closed.